Bring back legacy exception handling opcodes

trzeciak opened the pull request:
In this MR I bring back legacy exception handling opcodes, in reference to this report: https://github.com/bytecodealliance/wasm-tools/issues/1653.

@alexcrichton Two things:
- [x] I have prepared the simplest wasm file possible which uses all the deprecated opcodes (see `check_legacy_exception_handling_opcodes` test function),
could you please help me run the validator with this data?

- [x] Since this is behind the `legacy-exception-handling` feature, tests should be run with the `cargo test --all-features` flag, or I can add this feature to default list.
Given this, how should I do it to make sense?

alexcrichton commented on crates/wasmparser/Cargo.toml:
```diff
@@ -69,3 +70,8 @@ validate = [
+# A feature that enables legacy exception handling proposal (phase 1)
+# support this feature as long as all leading browsers also support it
+# https://github.com/WebAssembly/exception-handling/blob/main/proposals/exception-handling/legacy/Exceptions.md
+legacy-exception-handling = []
```
Ah I apologize I fear I may have misled you. Currently wasmparser doesn't have other Cargo features for wasm proposals so there's not a lot of infrastructure to support this and as you're discovered there's bugs and things to shake out. Given all that it's ok to remove this and have the support always be compiled-in-by-default.

What I moreso meant is that we'll want a new feature gate in `WasmFeatures` which is turned off at runtime by default but support for it is compiled in by default.

trzeciak replied:
Updated ✅

alexcrichton commented on crates/wasmparser/src/validator/operators.rs:
```diff
@@ -121,6 +121,28 @@ pub enum FrameKind {
+    /// # Note
+    ///
+    /// This belongs to the now deprecated Wasm exception handling proposal from Phase 1.
+    DeprecatedTry,
```
Mind calling this "Legacy" instead of "Deprecated"?

Also mind updating the docs here to modernize what it points to?

trzeciak replied:
Updated ✅

alexcrichton commented on crates/wasmparser/tests/legacy-exception.rs:
```diff
@@ -0,0 +1,26 @@
+      )";
+
+    // TODO: Use `_binary_data` or `_text_format_data`
+}
```
I'd recommend adding this tests as `tests/local/legacy-exceptions.wast` where you could add a variety of modules as necessary.

trzeciak replied:
I don't fully understand this, I will take a closer look at it and let you know or ask for help.

trzeciak replied:
Updated ✅

trzeciak commented:
@alexcrichton Okay, I changed 'cargo feature' to WasmFeatures and I added simple wat test file. I also changed some names and description, I think it's ready for review.

alexcrichton commented on crates/wasmparser/src/validator/operators.rs:
```diff
@@ -4124,6 +4128,115 @@ where
         self.push_operand(ValType::I32)
     }
+    fn visit_try(&mut self, mut ty: BlockType) -> Self::Output {
+        if !self.features.legacy_exceptions() {
```
Oh this is automatically checked via the macro above so you can skip the check on `self.features` here and below

trzeciak replied:
Hm.. then these tests fail:

[test-01.log](https://github.com/user-attachments/files/16195427/test-01.log)

trzeciak replied:
Maybe it is because now, to parse legacy exceptions we need to set all or two flags:
```sh
❯ cargo build
❯ ./target/release/wasm-tools validate -fexception-handling,legacy-exceptions tests/local/legacy-exceptions.wat
```

alexcrichton replied:
Ah for that you'll need to update [this exception list](https://github.com/bytecodealliance/wasm-tools/blob/021389c149dc7fa0c0bac97c0bd84ffe4986d889/tests/roundtrip.rs#L646-L663)

alexcrichton replied:
You additionally might need to [remove legacy-exceptions from the default list](https://github.com/bytecodealliance/wasm-tools/blob/021389c149dc7fa0c0bac97c0bd84ffe4986d889/tests/roundtrip.rs#L578-L582) and then add a [special case](https://github.com/bytecodealliance/wasm-tools/blob/021389c149dc7fa0c0bac97c0bd84ffe4986d889/tests/roundtrip.rs#L627-L629) where it is enabled.

trzeciak replied:
Should I also update https://github.com/bytecodealliance/wasm-tools/blob/021389c149dc7fa0c0bac97c0bd84ffe4986d889/crates/wasmparser/src/lib.rs#L161-L165 to this:
```
// Deprecated old instructions from the exceptions proposal
@legacy_exceptions Try { blockty: $crate::BlockType } => visit_try
@legacy_exceptions Catch { tag_index: u32 } => visit_catch
@legacy_exceptions Rethrow { relative_depth: u32 } => visit_rethrow
@legacy_exceptions Delegate { relative_depth: u32 } => visit_delegate
@legacy_exceptions CatchAll => visit_catch_all
```
?

alexcrichton replied:
Oh sorry yes definitely!

trzeciak replied:
After add `"legacy-exceptions.wat" => features.insert(WasmFeatures::LEGACY_EXCEPTIONS),` and other fixes, some tests still failed:
[test-02.log](https://github.com/user-attachments/files/16195973/test-02.log)

trzeciak replied:
To be clear, I pushed all changes now (tests not pass locally).

alexcrichton replied:
That's [this comment](https://github.com/bytecodealliance/wasm-tools/pull/1661#discussion_r1676024130), notably updating [this list](https://github.com/bytecodealliance/wasm-tools/blob/021389c149dc7fa0c0bac97c0bd84ffe4986d889/tests/roundtrip.rs#L646-L663)
